The job **`i686-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/14247305216/job/39931466818) [(plain)](https://github.com/rust-lang-ci/rust/commit/0b5c0be3fb76bda6680aa00750d046b647445629/checks/39931466818/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (51620/52673)
Updating files:  99% (52147/52673)
Updating files: 100% (52673/52673)
Updating files: 100% (52673/52673), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
##[endgroup]
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format=%H
0b5c0be3fb76bda6680aa00750d046b647445629
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.url=https://github.com/rust-lang/rust
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
failures:

---- [codegen] tests\codegen\align-struct.rs stdout ----

error: verification with 'FileCheck' failed
status: exit code: 1
command: PATH=";C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\x64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.43.34808\bin\HostX64\x64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.43.34808\bin\HostX64\x86;D:\a\rust\rust\build\i686-pc-windows-msvc\stage0-bootstrap-tools\i686-pc-windows-msvc\release\deps;D:\a\rust\rust\build\i686-pc-windows-msvc\stage0\bin;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Users\runneradmin\bin;D:\a\rust\rust\ninja;D:\a\rust\rust\sccache;C:\Program Files\MongoDB\Server\5.0\bin;C:\aliyun-cli;C:\vcpkg;C:\Program Files (x86)\NSIS;C:\tools\zstd;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\3.5.1\x64;C:\cabal\bin;C:\ghcup\bin;C:\mingw64\bin;C:\Program Files\dotnet;C:\Program Files\MySQL\MySQL Server 8.0\bin;C:\Program Files\R\R-4.4.2\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\SeleniumWebDrivers\ChromeDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\usr\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.21.13\x64\bin;C:\hostedtoolcache\windows\Python\3.9.13\x64\Scripts;C:\hostedtoolcache\windows\Python\3.9.13\x64;C:\hostedtoolcache\windows\Ruby\3.0.7\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.442-6\x64\bin;C:\Program Files\ImageMagick-7.1.1-Q16-HDRI;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\ProgramData\Chocolatey\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files\dotnet;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\TortoiseSVN\bin;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files\Microsoft SQL Server\150\Tools\Binn;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\WiX Toolset v3.14\bin;C:\Program Files\Microsoft SQL Server\130\DTS\Binn;C:\Program Files\Microsoft SQL Server\140\DTS\Binn;C:\Program Files\Microsoft SQL Server\150\DTS\Binn;C:\Program Files\Microsoft SQL Server\160\DTS\Binn;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\CMake\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.9.9\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.43.34808\bin\HostX64\x86" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\llvm\\build\\bin\\FileCheck.exe" "--input-file" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\test\\codegen\\align-struct\\align-struct.ll" "D:\\a\\rust\\rust\\tests\\codegen\\align-struct.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
D:\a\rust\rust\tests\codegen\align-struct.rs:40:12: error: CHECK: expected string not found in input
 // CHECK: {{%.*}} = load i32, ptr {{%.*}}, align 64
           ^
D:\a\rust\rust\build\i686-pc-windows-msvc\test\codegen\align-struct\align-struct.ll:18:33: note: scanning from here
define noundef i32 @align64_load(ptr noalias nocapture noundef readonly align 64 dereferenceable(64) %a) unnamed_addr #0 {
                                ^
D:\a\rust\rust\build\i686-pc-windows-msvc\test\codegen\align-struct\align-struct.ll:20:2: note: possible intended match here
 %_0 = load i32, ptr %a, align 4, !noundef !2
 ^

Input file: D:\a\rust\rust\build\i686-pc-windows-msvc\test\codegen\align-struct\align-struct.ll
Check file: D:\a\rust\rust\tests\codegen\align-struct.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'align_struct.2bf4ead79139451a-cgu.0' 
            2: source_filename = "align_struct.2bf4ead79139451a-cgu.0" 
            3: target datalayout = "e-m:x-p:32:32-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32-a:0:32-S32" 
            4: target triple = "i686-pc-windows-msvc" 
            5:  
            6: ; Function Attrs: uwtable 
            7: define void @align64(ptr dead_on_unwind noalias nocapture noundef writable sret([64 x i8]) align 64 dereferenceable(64) %_0, i32 noundef %i) unnamed_addr #0 { 
            8: start: 
            9:  %a64 = alloca [64 x i8], align 64 
           10:  call void @llvm.lifetime.start.p0(i64 64, ptr %a64) 
           11:  store i32 %i, ptr %a64, align 4 
           12:  call void @llvm.memcpy.p0.p0.i32(ptr align 64 %_0, ptr align 64 %a64, i32 64, i1 false) 
           13:  call void @llvm.lifetime.end.p0(i64 64, ptr %a64) 
           14:  ret void 
           15: } 
           16:  
           17: ; Function Attrs: uwtable 
           18: define noundef i32 @align64_load(ptr noalias nocapture noundef readonly align 64 dereferenceable(64) %a) unnamed_addr #0 { 
check:40'0                                     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           19: start: 
check:40'0     ~~~~~~~
           20:  %_0 = load i32, ptr %a, align 4, !noundef !2 
check:40'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:40'1      ?                                             possible intended match
           21:  ret i32 %_0 
check:40'0     ~~~~~~~~~~~~~
           22: } 
check:40'0     ~~
           23:  
check:40'0     ~
           24: ; Function Attrs: uwtable 
check:40'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
           25: define void @nested64(ptr dead_on_unwind noalias nocapture noundef writable sret([128 x i8]) align 64 dereferenceable(128) %_0, ptr noalias nocapture noundef readonly align 64 dereferenceable(64) %a, i32 noundef %b, i32 noundef %c, i8 noundef %d) unnamed_addr #0 { 
check:40'0     ~~~~~~~~~~~~~~~~~~~~~
           26: start: 
           27:  %_6 = alloca [64 x i8], align 64 
           28:  %n64 = alloca [128 x i8], align 64 
           29:  call void @llvm.lifetime.start.p0(i64 128, ptr %n64) 
           30:  call void @llvm.lifetime.start.p0(i64 64, ptr %_6) 
           31:  call void @llvm.memcpy.p0.p0.i32(ptr align 64 %_6, ptr align 64 %a, i32 64, i1 false) 
           32:  call void @llvm.memcpy.p0.p0.i32(ptr align 64 %n64, ptr align 64 %_6, i32 64, i1 false) 
           33:  %0 = getelementptr inbounds i8, ptr %n64, i32 64 
           34:  store i32 %b, ptr %0, align 4 
           35:  %1 = getelementptr inbounds i8, ptr %n64, i32 68 
           36:  store i32 %c, ptr %1, align 4 
           37:  %2 = getelementptr inbounds i8, ptr %n64, i32 72 
           38:  store i8 %d, ptr %2, align 4 
           39:  call void @llvm.lifetime.end.p0(i64 64, ptr %_6) 
           40:  call void @llvm.memcpy.p0.p0.i32(ptr align 64 %_0, ptr align 64 %n64, i32 128, i1 false) 
           41:  call void @llvm.lifetime.end.p0(i64 128, ptr %n64) 
           42:  ret void 
           43: } 
           44:  
           45: ; Function Attrs: uwtable 
           46: define { i32, i32 } @enum4(i32 noundef %a) unnamed_addr #0 { 
           47: start: 
           48:  %e4 = alloca [8 x i8], align 4 
           49:  call void @llvm.lifetime.start.p0(i64 8, ptr %e4) 
           50:  %0 = getelementptr inbounds i8, ptr %e4, i32 4 
           51:  store i32 %a, ptr %0, align 4 
           52:  store i32 0, ptr %e4, align 4 
           53:  %_0.0 = load i32, ptr %e4, align 4, !range !3, !noundef !2 
           54:  %1 = getelementptr inbounds i8, ptr %e4, i32 4 
           55:  %_0.1 = load i32, ptr %1, align 4, !noundef !2 
           56:  call void @llvm.lifetime.end.p0(i64 8, ptr %e4) 
           57:  %2 = insertvalue { i32, i32 } poison, i32 %_0.0, 0 
           58:  %3 = insertvalue { i32, i32 } %2, i32 %_0.1, 1 
           59:  ret { i32, i32 } %3 
           60: } 
           61:  
           62: ; Function Attrs: uwtable 
           63: define void @enum64(ptr dead_on_unwind noalias nocapture noundef writable sret([128 x i8]) align 64 dereferenceable(128) %_0, ptr noalias nocapture noundef readonly align 64 dereferenceable(64) %a) unnamed_addr #0 { 
           64: start: 
           65:  %_3 = alloca [64 x i8], align 64 
           66:  %e64 = alloca [128 x i8], align 64 
           67:  call void @llvm.lifetime.start.p0(i64 128, ptr %e64) 
           68:  call void @llvm.lifetime.start.p0(i64 64, ptr %_3) 
           69:  call void @llvm.memcpy.p0.p0.i32(ptr align 64 %_3, ptr align 64 %a, i32 64, i1 false) 
           70:  %0 = getelementptr inbounds i8, ptr %e64, i32 64 
           71:  call void @llvm.memcpy.p0.p0.i32(ptr align 64 %0, ptr align 64 %_3, i32 64, i1 false) 
           72:  store i32 0, ptr %e64, align 4 
           73:  call void @llvm.lifetime.end.p0(i64 64, ptr %_3) 
           74:  call void @llvm.memcpy.p0.p0.i32(ptr align 64 %_0, ptr align 64 %e64, i32 128, i1 false) 
           75:  call void @llvm.lifetime.end.p0(i64 128, ptr %e64) 
           76:  ret void 
           77: } 
           78:  
           79: ; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: readwrite) 
           80: declare void @llvm.memcpy.p0.p0.i32(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i32, i1 immarg) #1 
           81:  
           82: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
           83: declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #2 
           84:  
           85: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
           86: declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #2 
           87:  
           88: attributes #0 = { uwtable "target-cpu"="pentium4" } 
           89: attributes #1 = { nocallback nofree nounwind willreturn memory(argmem: readwrite) } 
           90: attributes #2 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) } 
           91:  
           92: !llvm.module.flags = !{!0} 
           93: !llvm.ident = !{!1} 
           94:  
           95: !0 = !{i32 8, !"PIC Level", i32 2} 
           96: !1 = !{!"rustc version 1.88.0-nightly (0b5c0be3f 2025-04-03)"} 
           97: !2 = !{} 
           98: !3 = !{i32 0, i32 2} 
>>>>>>
------------------------------------------


---- [codegen] tests\codegen\issues\issue-56927.rs stdout ----

error: verification with 'FileCheck' failed
status: exit code: 1
command: PATH=";C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\x64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.43.34808\bin\HostX64\x64;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.43.34808\bin\HostX64\x86;D:\a\rust\rust\build\i686-pc-windows-msvc\stage0-bootstrap-tools\i686-pc-windows-msvc\release\deps;D:\a\rust\rust\build\i686-pc-windows-msvc\stage0\bin;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Users\runneradmin\bin;D:\a\rust\rust\ninja;D:\a\rust\rust\sccache;C:\Program Files\MongoDB\Server\5.0\bin;C:\aliyun-cli;C:\vcpkg;C:\Program Files (x86)\NSIS;C:\tools\zstd;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\3.5.1\x64;C:\cabal\bin;C:\ghcup\bin;C:\mingw64\bin;C:\Program Files\dotnet;C:\Program Files\MySQL\MySQL Server 8.0\bin;C:\Program Files\R\R-4.4.2\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\SeleniumWebDrivers\ChromeDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\usr\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.21.13\x64\bin;C:\hostedtoolcache\windows\Python\3.9.13\x64\Scripts;C:\hostedtoolcache\windows\Python\3.9.13\x64;C:\hostedtoolcache\windows\Ruby\3.0.7\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.442-6\x64\bin;C:\Program Files\ImageMagick-7.1.1-Q16-HDRI;C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\ProgramData\Chocolatey\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\Program Files\dotnet;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\TortoiseSVN\bin;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files\Microsoft SQL Server\150\Tools\Binn;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\WiX Toolset v3.14\bin;C:\Program Files\Microsoft SQL Server\130\DTS\Binn;C:\Program Files\Microsoft SQL Server\140\DTS\Binn;C:\Program Files\Microsoft SQL Server\150\DTS\Binn;C:\Program Files\Microsoft SQL Server\160\DTS\Binn;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\CMake\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.9.9\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps;C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.43.34808\bin\HostX64\x86" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\llvm\\build\\bin\\FileCheck.exe" "--input-file" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\test\\codegen\\issues\\issue-56927\\issue-56927.ll" "D:\\a\\rust\\rust\\tests\\codegen\\issues\\issue-56927.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
D:\a\rust\rust\tests\codegen\issues\issue-56927.rs:11:11: error: CHECK: expected string not found in input
// CHECK: store i32 0, ptr %{{.+}}, align 16
          ^
D:\a\rust\rust\build\i686-pc-windows-msvc\test\codegen\issues\issue-56927\issue-56927.ll:11:19: note: scanning from here
define void @test1(ptr noalias noundef align 4 dereferenceable(16) %s) unnamed_addr #0 {
                  ^
D:\a\rust\rust\build\i686-pc-windows-msvc\test\codegen\issues\issue-56927\issue-56927.ll:14:2: note: possible intended match here
 store i32 0, ptr %0, align 4
 ^

Input file: D:\a\rust\rust\build\i686-pc-windows-msvc\test\codegen\issues\issue-56927\issue-56927.ll
Check file: D:\a\rust\rust\tests\codegen\issues\issue-56927.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'issue_56927.9d9fe2d6461d79bf-cgu.0' 
            2: source_filename = "issue_56927.9d9fe2d6461d79bf-cgu.0" 
            3: target datalayout = "e-m:x-p:32:32-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32-a:0:32-S32" 
            4: target triple = "i686-pc-windows-msvc" 
            5:  
            6: @alloc_da518201f2f3096fb01fb8522d2ee26c = private unnamed_addr constant [50 x i8] c"D:\\a\\rust\\rust\\tests\\codegen\\issues\\issue-56927.rs", align 1 
            7: @alloc_74b21464bd6bf852f742cf18994ed32b = private unnamed_addr constant <{ ptr, [12 x i8] }> <{ ptr @alloc_da518201f2f3096fb01fb8522d2ee26c, [12 x i8] c"2\00\00\00\1C\00\00\00\05\00\00\00" }>, align 4 
            8: @alloc_b64703bca8297e4012fef7725a74a74f = private unnamed_addr constant <{ ptr, [12 x i8] }> <{ ptr @alloc_da518201f2f3096fb01fb8522d2ee26c, [12 x i8] c"2\00\00\00#\00\00\00\05\00\00\00" }>, align 4 
            9:  
           10: ; Function Attrs: uwtable 
           11: define void @test1(ptr noalias noundef align 4 dereferenceable(16) %s) unnamed_addr #0 { 
check:11'0                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           12: start: 
check:11'0     ~~~~~~~
           13:  %0 = getelementptr inbounds nuw i32, ptr %s, i32 0 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14:  store i32 0, ptr %0, align 4 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:11'1      ?                             possible intended match
           15:  %1 = getelementptr inbounds nuw i32, ptr %s, i32 1 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16:  store i32 1, ptr %1, align 4 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           17:  %2 = getelementptr inbounds nuw i32, ptr %s, i32 2 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           18:  store i32 2, ptr %2, align 4 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           19:  %3 = getelementptr inbounds nuw i32, ptr %s, i32 3 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           20:  store i32 3, ptr %3, align 4 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21:  ret void 
check:11'0     ~~~~~~~~~~
           22: } 
check:11'0     ~~
           23:  
check:11'0     ~
           24: ; Function Attrs: uwtable 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
           25: define void @test2(ptr noalias noundef align 4 dereferenceable(16) %s) unnamed_addr #0 { 
check:11'0     ~~~~~~~~~~~~~~~~~~
           26: start: 
           27:  br i1 false, label %bb1, label %panic 
           28:  
           29: bb1: ; preds = %start 
           30:  %0 = getelementptr inbounds nuw i32, ptr %s, i32 1073741824 
           31:  store i32 4, ptr %0, align 4 
           32:  ret void 
           33:  
           34: panic: ; preds = %start 
           35: ; call core::panicking::panic_bounds_check 
           36:  call void @_ZN4core9panicking18panic_bounds_check17hcd20bb5be7796745E(i32 noundef 1073741824, i32 noundef 4, ptr noalias noundef readonly align 4 dereferenceable(16) @alloc_74b21464bd6bf852f742cf18994ed32b) #2 
           37:  unreachable 
           38: } 
           39:  
           40: ; Function Attrs: uwtable 
           41: define void @test3(ptr noalias noundef align 4 dereferenceable(16) %s, i32 noundef %i) unnamed_addr #0 { 
           42: start: 
           43:  %_3 = icmp ult i32 %i, 4 
           44:  br i1 %_3, label %bb1, label %panic 
           45:  
           46: bb1: ; preds = %start 
           47:  %0 = getelementptr inbounds nuw i32, ptr %s, i32 %i 
           48:  store i32 5, ptr %0, align 4 
           49:  ret void 
           50:  
           51: panic: ; preds = %start 
           52: ; call core::panicking::panic_bounds_check 
           53:  call void @_ZN4core9panicking18panic_bounds_check17hcd20bb5be7796745E(i32 noundef %i, i32 noundef 4, ptr noalias noundef readonly align 4 dereferenceable(16) @alloc_b64703bca8297e4012fef7725a74a74f) #2 
           54:  unreachable 
           55: } 
           56:  
           57: ; Function Attrs: uwtable 
           58: define void @test4(ptr noalias noundef align 4 dereferenceable(16) %s) unnamed_addr #0 { 
           59: start: 
           60:  br label %repeat_loop_header 
           61:  
           62: repeat_loop_header: ; preds = %repeat_loop_body, %start 
           63:  %0 = phi i32 [ 0, %start ], [ %3, %repeat_loop_body ] 
           64:  %1 = icmp ult i32 %0, 4 
           65:  br i1 %1, label %repeat_loop_body, label %repeat_loop_next 
           66:  
           67: repeat_loop_body: ; preds = %repeat_loop_header 
           68:  %2 = getelementptr inbounds nuw i32, ptr %s, i32 %0 
           69:  store i32 6, ptr %2, align 4 
           70:  %3 = add nuw i32 %0, 1 
           71:  br label %repeat_loop_header 
           72:  
           73: repeat_loop_next: ; preds = %repeat_loop_header 
           74:  ret void 
           75: } 
           76:  
           77: ; core::panicking::panic_bounds_check 
           78: ; Function Attrs: cold minsize noinline noreturn optsize uwtable 
           79: declare void @_ZN4core9panicking18panic_bounds_check17hcd20bb5be7796745E(i32 noundef, i32 noundef, ptr noalias noundef readonly align 4 dereferenceable(16)) unnamed_addr #1 
           80:  
           81: attributes #0 = { uwtable "target-cpu"="pentium4" } 
           82: attributes #1 = { cold minsize noinline noreturn optsize uwtable "target-cpu"="pentium4" } 
           83: attributes #2 = { noreturn } 
           84:  
           85: !llvm.module.flags = !{!0} 
           86: !llvm.ident = !{!1} 
           87:  
           88: !0 = !{i32 8, !"PIC Level", i32 2} 
           89: !1 = !{!"rustc version 1.88.0-nightly (0b5c0be3f 2025-04-03)"} 
>>>>>>
------------------------------------------



---
test result: FAILED. 623 passed; 2 failed; 215 ignored; 0 measured; 3 filtered out; finished in 23.40s

Some tests failed in compiletest suite=codegen mode=codegen host=i686-pc-windows-msvc target=i686-pc-windows-msvc
Build completed unsuccessfully in 1:40:28
make: *** [Makefile:113: ci-msvc-py] Error 1
  local time: Thu Apr  3 17:38:42 CUT 2025
  network time: Thu, 03 Apr 2025 17:38:42 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

--- a/compiler/rustc_codegen_llvm/src/builder.rs
+++ b/compiler/rustc_codegen_llvm/src/builder.rs
@@ -594,6 +594,7 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
     fn load(&mut self, ty: &'ll Type, ptr: &'ll Value, align: Align) -> &'ll Value {
         unsafe {
             let load = llvm::LLVMBuildLoad2(self.llbuilder, ty, ptr, UNNAMED);
+            let align = align.min(self.cx().tcx.sess.target.max_reliable_alignment());
             llvm::LLVMSetAlignment(load, align.bytes() as c_uint);
             load
         }
@@ -807,6 +808,7 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
         assert_eq!(self.cx.type_kind(self.cx.val_ty(ptr)), TypeKind::Pointer);
         unsafe {
             let store = llvm::LLVMBuildStore(self.llbuilder, val, ptr);
+            let align = align.min(self.cx().tcx.sess.target.max_reliable_alignment());
             let align =
                 if flags.contains(MemFlags::UNALIGNED) { 1 } else { align.bytes() as c_uint };
             llvm::LLVMSetAlignment(store, align);

--- a/compiler/rustc_mir_transform/src/check_alignment.rs
+++ b/compiler/rustc_mir_transform/src/check_alignment.rs
@@ -1,3 +1,4 @@
+use rustc_abi::Align;
 use rustc_index::IndexVec;
 use rustc_middle::mir::interpret::Scalar;
 use rustc_middle::mir::visit::PlaceContext;
@@ -11,8 +12,7 @@ pub(super) struct CheckAlignment;
 
 impl<'tcx> crate::MirPass<'tcx> for CheckAlignment {
     fn is_enabled(&self, sess: &Session) -> bool {
-        // FIXME(#112480) MSVC and rustc disagree on minimum stack alignment on x86 Windows
-        if sess.target.llvm_target == "i686-pc-windows-msvc" {
+        if sess.target.max_reliable_alignment() < Align::MAX {
             return false;
         }
         sess.ub_checks()

--- a/compiler/rustc_mir_transform/src/check_alignment.rs
+++ b/compiler/rustc_mir_transform/src/check_alignment.rs
@@ -12,9 +12,6 @@ pub(super) struct CheckAlignment;
 
 impl<'tcx> crate::MirPass<'tcx> for CheckAlignment {
     fn is_enabled(&self, sess: &Session) -> bool {
-        if sess.target.max_reliable_alignment() < Align::MAX {
-            return false;
-        }
         sess.ub_checks()
     }
 
@@ -86,6 +83,33 @@ fn insert_alignment_check<'tcx>(
             Rvalue::BinaryOp(BinOp::Sub, Box::new((Operand::Copy(alignment), one))),
         ))),
     });
+
+    // If this target does not have reliable alignment, further limit the mask by anding it with
+    // the mask for the highest reliable alignment.
+    #[allow(irrefutable_let_patterns)]
+    if let max_align = tcx.sess.target.max_reliable_alignment()
+        && max_align < Align::MAX
+    {
+        let max_mask = max_align.bytes() - 1;
+        let max_mask = Operand::Constant(Box::new(ConstOperand {
+            span: source_info.span,
+            user_ty: None,
+            const_: Const::Val(
+                ConstValue::Scalar(Scalar::from_target_usize(max_mask, &tcx)),
+                tcx.types.usize,
+            ),
+        }));
+        stmts.push(Statement {
+            source_info,
+            kind: StatementKind::Assign(Box::new((
+                alignment_mask,
+                Rvalue::BinaryOp(
+                    BinOp::BitAnd,
+                    Box::new((Operand::Copy(alignment_mask), max_mask)),
+                ),
+            ))),
+        });
+    }
 
     // BitAnd the alignment mask with the pointer
     let alignment_bits =

--- a/compiler/rustc_target/src/callconv/mod.rs
+++ b/compiler/rustc_target/src/callconv/mod.rs
@@ -144,6 +144,7 @@ pub struct ArgAttributes {
     /// (corresponding to LLVM's dereferenceable_or_null attributes, i.e., it is okay for this to be
     /// set on a null pointer, but all non-null pointers must be dereferenceable).
     pub pointee_size: Size,
+    /// The minimum alignment of the pointee, if any.
     pub pointee_align: Option<Align>,
 }
 

--- a/compiler/rustc_target/src/spec/mod.rs
+++ b/compiler/rustc_target/src/spec/mod.rs
@@ -42,7 +42,9 @@ use std::path::{Path, PathBuf};
 use std::str::FromStr;
 use std::{fmt, io};
 
-use rustc_abi::{Endian, ExternAbi, Integer, Size, TargetDataLayout, TargetDataLayoutErrors};
+use rustc_abi::{
+    Align, Endian, ExternAbi, Integer, Size, TargetDataLayout, TargetDataLayoutErrors,
+};
 use rustc_data_structures::fx::{FxHashSet, FxIndexSet};
 use rustc_fs_util::try_canonicalize;
 use rustc_macros::{Decodable, Encodable, HashStable_Generic};
@@ -3597,6 +3599,25 @@ impl Target {
             // Unsupported architecture.
             _ => return None,
         })
+    }
+
+    /// Returns whether this target is known to have unreliable alignment:
+    /// native C code for the target fails to align some data to the degree
+    /// required by the C standard. We can't *really* do anything about that
+    /// since unsafe Rust code may assume alignment any time, but we can at least
+    /// inhibit some optimizations, and we suppress the alignment checks that
+    /// would detect this unsoundness.
+    ///
+    /// Every target that returns less than `Align::MAX` here is still has a soundness bug.
+    pub fn max_reliable_alignment(&self) -> Align {
+        // FIXME(#112480) MSVC on x86-32 is unsound and fails to properly align many types with
+        // more-than-4-byte-alignment on the stack. This makes alignments larger than 4 generally
+        // unreliable on 32bit Windows.
+        if self.is_like_windows && self.arch == "x86" {
+            Align::from_bytes(4).unwrap()
+        } else {
+            Align::MAX
+        }
     }
 }
 

--- a/compiler/rustc_ty_utils/src/abi.rs
+++ b/compiler/rustc_ty_utils/src/abi.rs
@@ -347,7 +347,8 @@ fn adjust_for_rust_scalar<'tcx>(
             None
         };
         if let Some(kind) = kind {
-            attrs.pointee_align = Some(pointee.align);
+            attrs.pointee_align =
+                Some(pointee.align.min(cx.tcx().sess.target.max_reliable_alignment()));
 
             // `Box` are not necessarily dereferenceable for the entire duration of the function as
             // they can be deallocated at any time. Same for non-frozen shared references (see

--- a/src/doc/rustc/src/platform-support.md
+++ b/src/doc/rustc/src/platform-support.md
@@ -34,7 +34,7 @@ target | notes
 -------|-------
 [`aarch64-apple-darwin`](platform-support/apple-darwin.md) | ARM64 macOS (11.0+, Big Sur+)
 `aarch64-unknown-linux-gnu` | ARM64 Linux (kernel 4.1, glibc 2.17+)
-`i686-pc-windows-msvc` | 32-bit MSVC (Windows 10+, Windows Server 2016+, Pentium 4) [^x86_32-floats-return-ABI]
+`i686-pc-windows-msvc` | 32-bit MSVC (Windows 10+, Windows Server 2016+, Pentium 4) [^x86_32-floats-return-ABI] [^win32-msvc-alignment]
 `i686-unknown-linux-gnu` | 32-bit Linux (kernel 3.2+, glibc 2.17+, Pentium 4) [^x86_32-floats-return-ABI]
 [`x86_64-apple-darwin`](platform-support/apple-darwin.md) | 64-bit macOS (10.12+, Sierra+)
 [`x86_64-pc-windows-gnu`](platform-support/windows-gnu.md) | 64-bit MinGW (Windows 10+, Windows Server 2016+)
@@ -42,6 +42,8 @@ target | notes
 `x86_64-unknown-linux-gnu` | 64-bit Linux (kernel 3.2+, glibc 2.17+)
 
 [^x86_32-floats-return-ABI]: Due to limitations of the C ABI, floating-point support on `i686` targets is non-compliant: floating-point return values are passed via an x87 register, so NaN payload bits can be lost. Functions with the default Rust ABI are not affected. See [issue #115567][x86-32-float-return-issue].
+
+[^win32-msvc-alignment]: Due to non-standard behavior of MSVC, native C code on this target can cause types with an alignment of more than 4 bytes to be incorrectly aligned to only 4 bytes (this affects, e.g., `u64` and `i64`). Rust applies some mitigations to reduce the impact of this issue, but this can still cause unsoundness due to unsafe code that (correctly) assumes that references are always properly aligned. See [issue #112480](https://github.com/rust-lang/rust/issues/112480).
 
 [77071]: https://github.com/rust-lang/rust/issues/77071
 [x86-32-float-return-issue]: https://github.com/rust-lang/rust/issues/115567
@@ -95,7 +97,7 @@ target | notes
 [`armv7-unknown-linux-ohos`](platform-support/openharmony.md) | Armv7-A OpenHarmony
 [`loongarch64-unknown-linux-gnu`](platform-support/loongarch-linux.md) | LoongArch64 Linux, LP64D ABI (kernel 5.19, glibc 2.36)
 [`loongarch64-unknown-linux-musl`](platform-support/loongarch-linux.md) | LoongArch64 Linux, LP64D ABI (kernel 5.19, musl 1.2.5)
-[`i686-pc-windows-gnu`](platform-support/windows-gnu.md) | 32-bit MinGW (Windows 10+, Windows Server 2016+, Pentium 4) [^x86_32-floats-return-ABI]
+[`i686-pc-windows-gnu`](platform-support/windows-gnu.md) | 32-bit MinGW (Windows 10+, Windows Server 2016+, Pentium 4) [^x86_32-floats-return-ABI] [^win32-msvc-alignment]
 `powerpc-unknown-linux-gnu` | PowerPC Linux (kernel 3.2, glibc 2.17)
 `powerpc64-unknown-linux-gnu` | PPC64 Linux (kernel 3.2, glibc 2.17)
 [`powerpc64le-unknown-linux-gnu`](platform-support/powerpc64le-unknown-linux-gnu.md) | PPC64LE Linux (kernel 3.10, glibc 2.17)
@@ -169,7 +171,7 @@ target | std | notes
 [`i686-pc-windows-gnullvm`](platform-support/windows-gnullvm.md) | ✓ | 32-bit x86 MinGW (Windows 10+, Pentium 4), LLVM ABI [^x86_32-floats-return-ABI]
 [`i686-unknown-freebsd`](platform-support/freebsd.md) | ✓ | 32-bit x86 FreeBSD (Pentium 4) [^x86_32-floats-return-ABI]
 `i686-unknown-linux-musl` | ✓ | 32-bit Linux with musl 1.2.3 (Pentium 4) [^x86_32-floats-return-ABI]
-[`i686-unknown-uefi`](platform-support/unknown-uefi.md) | ? | 32-bit UEFI (Pentium 4, softfloat)
+[`i686-unknown-uefi`](platform-support/unknown-uefi.md) | ? | 32-bit UEFI (Pentium 4, softfloat) [^win32-msvc-alignment]
 [`loongarch64-unknown-none`](platform-support/loongarch-none.md) | * | LoongArch64 Bare-metal (LP64D ABI)
 [`loongarch64-unknown-none-softfloat`](platform-support/loongarch-none.md) | * | LoongArch64 Bare-metal (LP64S ABI)
 [`nvptx64-nvidia-cuda`](platform-support/nvptx64-nvidia-cuda.md) | * | --emit=asm generates PTX code that [runs on NVIDIA GPUs]
@@ -317,9 +319,9 @@ target | std | host | notes
 [`i686-unknown-netbsd`](platform-support/netbsd.md) | ✓ | ✓ | NetBSD/i386 (Pentium 4) [^x86_32-floats-return-ABI]
 [`i686-unknown-openbsd`](platform-support/openbsd.md) | ✓ | ✓ | 32-bit OpenBSD (Pentium 4) [^x86_32-floats-return-ABI]
 `i686-uwp-windows-gnu` | ✓ |  | [^x86_32-floats-return-ABI]
-[`i686-uwp-windows-msvc`](platform-support/uwp-windows-msvc.md) | ✓ |  | [^x86_32-floats-return-ABI]
+[`i686-uwp-windows-msvc`](platform-support/uwp-windows-msvc.md) | ✓ |  | [^x86_32-floats-return-ABI] [^win32-msvc-alignment]
 [`i686-win7-windows-gnu`](platform-support/win7-windows-gnu.md) | ✓ |   | 32-bit Windows 7 support [^x86_32-floats-return-ABI]
-[`i686-win7-windows-msvc`](platform-support/win7-windows-msvc.md) | ✓ |   | 32-bit Windows 7 support [^x86_32-floats-return-ABI]
+[`i686-win7-windows-msvc`](platform-support/win7-windows-msvc.md) | ✓ |   | 32-bit Windows 7 support [^x86_32-floats-return-ABI] [^win32-msvc-alignment]
 [`i686-wrs-vxworks`](platform-support/vxworks.md) | ✓ |  | [^x86_32-floats-return-ABI]
 [`loongarch64-unknown-linux-ohos`](platform-support/openharmony.md) | ✓ |   | LoongArch64 OpenHarmony
 [`m68k-unknown-linux-gnu`](platform-support/m68k-unknown-linux-gnu.md) | ? |  | Motorola 680x0 Linux

--- a/tests/codegen/align-struct.rs
+++ b/tests/codegen/align-struct.rs
@@ -1,5 +1,7 @@
 //@ compile-flags: -C no-prepopulate-passes -Z mir-opt-level=0
-//
+// 32bit MSVC does not align things properly so we suppress high alignment annotations (#112480)
+//@ ignore-i686-pc-windows-msvc
+//@ ignore-i686-pc-windows-gnu
 
 #![crate_type = "lib"]
 

--- a/tests/codegen/issues/issue-56927.rs
+++ b/tests/codegen/issues/issue-56927.rs
@@ -1,4 +1,7 @@
 //@ compile-flags: -C no-prepopulate-passes
+// 32bit MSVC does not align things properly so we suppress high alignment annotations (#112480)
+//@ ignore-i686-pc-windows-msvc
+//@ ignore-i686-pc-windows-gnu
 
 #![crate_type = "rlib"]
 

